src/poetry/console/commands/build.py

```diff
@@ -1,18 +1,150 @@
 from __future__ import annotations
 
+import dataclasses
+
 from pathlib import Path
 from typing import TYPE_CHECKING
+from typing import Any
 from typing import ClassVar
+from typing import Literal
 
 from cleo.helpers import option
+from poetry.core.constraints.version import Version
 
 from poetry.console.commands.env_command import EnvCommand
-from poetry.utils.env import build_environment
+from poetry.masonry.builders import BUILD_FORMATS
+from poetry.utils._compat import metadata
 from poetry.utils.helpers import remove_directory
+from poetry.utils.isolated_build import isolated_builder
 
 
 if TYPE_CHECKING:
+    from collections.abc import Callable
+
     from cleo.io.inputs.option import Option
+    from cleo.io.io import IO
+
+    from poetry.poetry import Poetry
+    from poetry.utils.env import Env
+
+DistributionType = Literal["sdist", "wheel"]
+
+
+@dataclasses.dataclass(frozen=True)
+class BuildOptions:
+    clean: bool
+    formats: list[DistributionType]
+    output: str
+    config_settings: dict[str, Any] = dataclasses.field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        for fmt in self.formats:
+            if fmt not in BUILD_FORMATS:
+                raise ValueError(f"Invalid format: {fmt}")
+
+
+class BuildHandler:
+    def __init__(self, poetry: Poetry, env: Env, io: IO) -> None:
+        self.poetry = poetry
+        self.env = env
+        self.io = io
+
+    def _build(
+        self,
+        fmt: DistributionType,
+        executable: Path,
+        target_dir: Path,
+        config_settings: dict[str, Any],
+    ) -> None:
+        builder = BUILD_FORMATS[fmt]
+
+        builder(
+            self.poetry,
+            executable=executable,
+            config_settings=config_settings,
+        ).build(target_dir)
+
+    def _isolated_build(
+        self,
+        fmt: DistributionType,
+        executable: Path,
+        target_dir: Path,
+        config_settings: dict[str, Any],
+    ) -> None:
+        with isolated_builder(
+            source=self.poetry.file.path.parent,
+            distribution=fmt,
+            python_executable=executable,
+        ) as builder:
+            builder.build(fmt, target_dir, config_settings=config_settings)
+
+    def _requires_isolated_build(self) -> bool:
+        """
+        Determines if an isolated build is required.
+
+        An isolated build is required if:
+        - The package has a build script.
+        - There are multiple build system dependencies.
+        - The build dependency is not `poetry-core`.
+        - The installed `poetry-core` version does not satisfy the build dependency constraints.
+        - The build dependency has a source type (e.g. is a VcsDependency).
+
+        :returns: True if an isolated build is required, False otherwise.
+        """
+        if (
+            self.poetry.package.build_script
```

Comment:
@finswimmer can you do a followup pull request to use the new mechanism even if we have a build script? I want to keep that in a separate change so we can revert if required later.

Reply:
Not sure if understand what you want me to do :thinking:

Reply:
Nevermind I'm an idiot, had my wires crossed.